### PR TITLE
ROX-25492, ROX-25493: Configure Policy-as-code in operator

### PR DESCRIPTION
**Pull request by kylape**
### CRD Installation in Operator

For operator-based installs of ACS, CRDs are currently installed by OLM.  While the `SecurityPolicy` CRD is unique in that it is actually used by the StackRox application, OLM is still the most intuitive method of installation.

This approach takes two steps:

1. Copy the CRD from config-controller over to the operator bundle
2. Add an ownership reference to the `SecurityPolicy` CRD in the ACS operator CSV

An alternative approach was considered where the `operator-sdk` CLI was "tricked" into generating the `SecurityPolicy` CRD, but this required copying over the go structs for the `SecurityPolicy` CRD, which seemed like much more work overall.

### `configAsCode` toggle in Helm Chart

This adds a Helm chart parameter to toggle config-as-code.  Defaults to `true`.

The name `configAsCode` is used instead of `policyAsCode` to make the API future-proof in case we add other config-related CRDs down the road.

Example:

```
configAsCode:
  enabled: false
```

### Make configAsCode configurable in operator

Exposing the above Helm chart toggle `configAsCode.enabled` in the `Central` CRD, defaulting to `Enabled`.

### Apply CRD in deploy script

Manifest installation will not include the CRD, so we are adding it "by hand" in the deploy script before either Helm or manifest installation methods have been executred.

### CODE REVIEW

File: image/templates/helm/stackrox-central/internal/defaults.yaml.htpl [hunk 1]
```diff
@@ -314,6 +314,10 @@ defaults:
         tag: [< required "" .ScannerV4ImageTag >]
     exposeMonitoring: false
 
+  configAsCode:
+    enabled: true
```
Comment (clickboo) [inline image/templates/helm/stackrox-central/internal/defaults.yaml.htpl:318]: enabled should technically be false till we turn on feature flag by default. If we for any reason make the decision to not ship with feature flag enabled by default this release, we should not be deploying config controller?

File: image/templates/helm/stackrox-central/templates/01-central-13-deployment.yaml.htpl [hunk 2]
```diff
@@ -142,6 +142,10 @@ spec:
           value: "true"
         {{- end }}
         [<- end >]
+        [< if .FeatureFlags.ROX_POLICY_AS_CODE >]
```
Comment (clickboo) [inline image/templates/helm/stackrox-central/templates/01-central-13-deployment.yaml.htpl:145]: This feature flag has no meaning if we don't check for it in the config controller - we should.

File: image/templates/helm/stackrox-central/values.yaml.htpl [hunk 3]
```diff
@@ -538,7 +538,12 @@ central:
 #    # sufficient and manual override is required.
 #    configOverride: null
 #
+[< if .FeatureFlags.ROX_POLICY_AS_CODE >]
```
Comment (clickboo) [inline image/templates/helm/stackrox-central/values.yaml.htpl:541]: Interesting there is a feature flag check here before setting the values as true which is the correct thing to do, but there is the config yaml where we don't check feature flag. We should for both correctness and consistency right?

File: image/templates/helm/stackrox-central/templates/01-central-13-deployment.yaml.htpl (hunk 2 above)
Comment (kylape) [reply image/templates/helm/stackrox-central/templates/01-central-13-deployment.yaml.htpl:145]: One way to describe what this is doing in plain English: "If the policy-as-code feature flag is enabled when the Helm template is generated, add the `ROX_POLICY_AS_CODE` flag to the Central deployment, with the value being read from the `configAsCode.enabled` field in the values file.

The thinking is that the env var shouldn't be added at all if the Helm chart wasn't generated with policy-as-code enabled since there's no way the config-controller would ever get deployed (the config-controller template files are excluded).  That said, I admit it's kind of confusing that the same env var (`ROX_POLICY_AS_CODE`) is used at two different levels here.

File: image/templates/helm/stackrox-central/values.yaml.htpl (hunk 3 above)
Comment (kylape) [reply image/templates/helm/stackrox-central/values.yaml.htpl:541]: Are you referring to `config-share.yaml`?

File: image/templates/helm/stackrox-central/internal/defaults.yaml.htpl (hunk 1 above)
Comment (kylape) [reply image/templates/helm/stackrox-central/internal/defaults.yaml.htpl:318]: Should we remove the feature flag now that we can disable policy-as-code via helm chart values?  The main thing we lose is protection of master from silly Helm chart template generation goofs (which we recently had!).

File: image/templates/helm/stackrox-central/templates/01-central-13-deployment.yaml.htpl (hunk 2 above)
Comment (mclasmeier) [inline image/templates/helm/stackrox-central/templates/01-central-13-deployment.yaml.htpl:145]: Please use `[<- ... >]` for this kind of line-based control-flow. Reasons are explaines in https://github.com/stackrox/stackrox/blob/fd6f4c86f36f2b4d6ebd3a56ba4bdb35fbabf177/image/templates/CHANGING_CHARTS.md#whitespace-elimination and apply equally to the meta-templating.

File: image/templates/helm/stackrox-central/templates/02-config-controller-02-deployment.yaml [hunk 4]
```diff
@@ -44,6 +45,7 @@ spec:
           valueFrom:
             fieldRef:
               fieldPath: metadata.namespace
+        {{- include "srox.envVars" (list . "deployment" "config-controller" "config-controller") | nindent 8 }}
```
Comment (mclasmeier) [inline image/templates/helm/stackrox-central/templates/02-config-controller-02-deployment.yaml:48]: The fourth parameter needs to be the container name.

File: image/templates/helm/stackrox-central/values.yaml.htpl (hunk 3 above)
Comment (mclasmeier) [reply image/templates/helm/stackrox-central/values.yaml.htpl:541]: `[<- .... >]`

File: image/templates/helm/stackrox-central/values.yaml.htpl [hunk 5]
```diff
@@ -538,7 +538,12 @@ central:
...
+#configAsCode:
+#  enabled: true
+#  installCRDs: true
```
Comment (mclasmeier) [inline image/templates/helm/stackrox-central/values.yaml.htpl:544]: Looking at it from a user perspective, I am wondering if we should keep `installCRDs` somewhat internal (for example, for usage by the Operator). I mean, what would be the use-cases where installing the CRDs should be disabled?

File: operator/Makefile [hunk 6]
```diff
@@ -434,6 +434,7 @@ bundle-post-process: test-bundle-helpers operator-sdk ## Post-process CSV file t
 	mkdir -p build/
 	rm -rf build/bundle
 	cp -a bundle build/
+	cp ../config-controller/config/crd/bases/*.yaml build/bundle/manifests/ ;\
```
Comment (mclasmeier) [inline operator/Makefile:437]:
```suggestion
	cp ../config-controller/config/crd/bases/*.yaml build/bundle/manifests/
```

This seems independent of the commands below.

File: operator/api/v1alpha1/central_types.go [hunk 7]
```diff
@@ -611,6 +615,10 @@ const (
 	ScannerV4ComponentDisabled ScannerV4ComponentPolicy = "Disabled"
 )
 
+type ConfigAsCodeSpec struct {
+	Enabled *bool `json:"enabled,omitempty"`
```
Comment (mclasmeier) [inline operator/api/v1alpha1/central_types.go:619]: General remark: I think we should try to aim for consistency -- at least for new features. In the recent past I think there was a tendency to use the reversed logic and add `disabled` switches (e.g. for Scanner V4) instead of `enabled` switches. WDYT?

File: operator/internal/values/translation/translation.go [hunk 8]
```diff
@@ -416,3 +416,12 @@ func GetGlobalNetwork(s *platform.GlobalNetworkSpec) *ValuesBuilder {
 	}
 	return &sv
 }
+
+// GetConfigAsCode converts *platform.GetConfigAsCodeSpec into *ValuesBuilder
+func GetConfigAsCode(c *platform.ConfigAsCodeSpec) *ValuesBuilder {
+	sv := NewValuesBuilder()
+	if c.Enabled != nil {
```
Comment (mclasmeier) [inline operator/internal/values/translation/translation.go:423]: Just want to clarify: the behavior here is that when `configAsCode` is not configured in the CR, then the Helm chart defaults will end up being used, right?

File: operator/bundle_helpers/patch-csv.py [hunk 9]
```diff
@@ -85,6 +85,29 @@ def patch_csv(csv_doc, version, operator_image, first_version, no_related_images
...
+            "name": "",
+            "version": "v1",
+        }],
+    }
+
+    csv_doc["spec"]["customresourcedefinitions"]["owned"].append(policy_crd)
+
+    # Apply policy-as-code feature flag to operator
+    env = csv_doc["spec"]["install"]["spec"]["deployments"][0]["spec"]["template"]["spec"]["containers"][0]["env"]
+    env.append({
```
Comment (mclasmeier) [inline operator/bundle_helpers/patch-csv.py:106]: I assume this is currently needed, because the feature flag is off by default -- correct?

I think we should keep `patch-csv.py` as simple as possible, hence I am wondering, if you need a static setting in the CSV, why don't you simply add it to the CSV directly? Am I missing something here?

**Review by mclasmeier — CHANGES_REQUESTED** [review]

File: image/templates/helm/stackrox-central/internal/config-shape.yaml [hunk 10]
```diff
@@ -164,6 +170,9 @@ allowNonstandardNamespace: null # bool
 allowNonstandardReleaseName: null # bool
 allowUnsupportedHelmVersion: null # bool
 enableOpenShiftMonitoring: null # bool
+configAsCode:
```
Comment (SimonBaeumer) [inline image/templates/helm/stackrox-central/internal/config-shape.yaml:173]: Central has already a `declarativeConfig` section with the same intention of providing configs as code. I found the `declarativeConfig` not optimal but customers are already using it.
Maybe putting there on a feature basis, e.g. `policiesAsCode.enabled`? Eventually more could be added there.

https://github.com/stackrox/stackrox/blob/7d176aaff7da22a5e4ff306362a2dd2ae88d7e72/image/templates/helm/stackrox-central/internal/config-shape.yaml#L74-L77

**Review by SimonBaeumer — COMMENTED** [review]
Ter

File: image/templates/helm/stackrox-central/internal/config-shape.yaml [hunk 11]
```diff
@@ -129,6 +129,12 @@ customize:
       podLabels: {} # dict
       podAnnotations: {} # dict
       envVars: {} # dict
+  config-controller:
```
[the remark this answers is not in the document]
Comment (kylape) [reply image/templates/helm/stackrox-central/internal/config-shape.yaml:132]: Looks like this matches the name of the deployment, which follows the k8s convention of using dashes in the name.  The other deployments names in this section of the YAML follow the same convention.

[the remark this answers is not in the document]
Comment (mclasmeier) [reply image/templates/helm/stackrox-central/internal/config-shape.yaml:132]: you are totally right, sorry.

File: image/templates/helm/stackrox-central/values.yaml.htpl (hunk 5 above)
Comment (kylape) [reply image/templates/helm/stackrox-central/values.yaml.htpl:544]: The current plan is to install the `SecurityPolicy` CRD even when config-as-code is disabled.  The reasoning is that we believe the typical use case for disabling config-as-code will be to turn it off after it was enabled by default as opposed to disabling it up front.  In that use case, we want to keep the CRD installed even when policy-as-code is disabled so that any policy CRs aren't lost.

Given that the CRD will be installed even when policy-as-code is disabled, there needs to be a way for users to disable installing the CRD since it is the only resource that requires elevated privileges and is installed at the cluster scope.

Open to discuss any of the reasoning laid out above.

File: operator/api/v1alpha1/central_types.go (hunk 7 above)
Comment (kylape) [reply operator/api/v1alpha1/central_types.go:619]: Ha, I did consider whether it should be `enabled` or `disabled`.  I went with `enabled` because it's my preference and also because it wasn't clear to me looking at the values template that there was a consistent pattern.

Taking a look again, I see these uses:

`disable`: scanner, scannerv4, autoscaling
`enabled`: monitoring, telemetry, exposure

That said, I'm okay to change it if you want to make `disabled` the established pattern.

File: operator/internal/values/translation/translation.go (hunk 8 above)
Comment (kylape) [reply operator/internal/values/translation/translation.go:423]: If `configAsCode` is not set in the CR, then the values applied to the helm chart will effectively be `configAsCode.enabled=true` (coming from Helm chart defaults) and `configAsCode.installCRD=false` (coming from `operator/internal/central/values/translation/base-values.yaml`)

File: operator/bundle_helpers/patch-csv.py (hunk 9 above)
Comment (kylape) [reply operator/bundle_helpers/patch-csv.py:106]: I believe [the CSV gets regenerated from scratch](https://github.com/stackrox/stackrox/blob/35514dd94dd78a62ec7080a084afb383ea06b121/operator/Makefile#L402-L404) during `make bundle`.

File: image/templates/helm/stackrox-central/internal/config-shape.yaml (hunk 10 above)
Comment (kylape) [reply image/templates/helm/stackrox-central/internal/config-shape.yaml:173]: I'm torn on this suggestion.  This (configAsCode) is a distinct feature from declarativeConfig, but both features could certainly be described as "declarative config".  It might be confusing to customers to have both a `declarativeConfig` and `configAsCode` section, but at the same time it might be confusing to mash two distinct features under one section in the values.yaml.

Maybe a different name?

```
declarativeConfig:
  kubeNative:
    policies: true
```

This mirrors the structure of declarativeConfig.

Another consideration is whether to put configAsCode config under the `central` key or not.  I was leaning towards not since it's a separate deployment, following the pattern of scanner.

File: operator/Makefile (hunk 6 above)
Comment (kylape) [reply operator/Makefile]: cadbc63

File: image/templates/helm/stackrox-central/templates/02-config-controller-02-deployment.yaml (hunk 4 above)
Comment (kylape) [reply image/templates/helm/stackrox-central/templates/02-config-controller-02-deployment.yaml]: 700c37f

File: operator/bundle_helpers/patch-csv.py (hunk 9 above)
Comment (mclasmeier) [reply operator/bundle_helpers/patch-csv.py:106]: It gets patched. See https://github.com/stackrox/stackrox/blob/35514dd94dd78a62ec7080a084afb383ea06b121/operator/Makefile#L458

File: image/templates/helm/stackrox-central/internal/config-shape.yaml (hunk 10 above)
Comment (SimonBaeumer) [reply image/templates/helm/stackrox-central/internal/config-shape.yaml:173]: Hm, I agree, declarative config feels not right too.
Is a `policyAsCode` config necessary? Could it just be under `config-controller.enabled` and enables automtically all CRD related stuff? I don't understand why the split between these configs was done.

File: operator/bundle_helpers/patch-csv.py (hunk 9 above)
Comment (kylape) [reply operator/bundle_helpers/patch-csv.py:106]: Yes, and I am modifying the script that does the patching to add the feature flag.  Perhpas it would be better if I added a more generic feature to the script?  Perhaps adding an `--add-env-var` to allow arbitrary environment variables to be set on the operator's deployment.

File: image/templates/helm/stackrox-central/internal/config-shape.yaml (hunk 10 above)
Comment (kylape) [reply image/templates/helm/stackrox-central/internal/config-shape.yaml:173]: I explained why the CRD has a separate toggle flag [here](https://github.com/stackrox/stackrox/pull/12799#discussion_r1782051317).  For now at least, disabling policy as code also removes a feature flag env var from the central deployment which disables the k8s service account m2m config.

File: image/templates/helm/stackrox-central/internal/defaults.yaml.htpl (hunk 1 above)
Comment (kylape) [reply image/templates/helm/stackrox-central/internal/defaults.yaml.htpl:318]: @clickboo I know we talked about this last week, and I think you ended up being _okay_ about removing the feature flag since we will still have a way to disable the deployment through the helm chart.  Is that correct?

File: operator/bundle_helpers/patch-csv.py (hunk 9 above)
Comment (mclasmeier) [reply operator/bundle_helpers/patch-csv.py:109]: I am saying, why changing the *script* instead of just modifying the input of the script, namely `bundle/manifests/rhacs-operator.clusterserviceversion.yaml`?
After all you are just adding a static setting.
The script should be only used for dynamic settings, e.g. image tags/digests.
Otherwise we could end up in tricky situtations, because it is not clear which parts of the resulting CSV are coming from the source yaml and which ones are added via the script.

Comment (kylape) [reply operator/bundle_helpers/patch-csv.py:109]: `bundle/manifests/rhacs-operator.clusterserviceversion.yaml` is deleted and re-generated from within the `bundle` target:

https://github.com/stackrox/stackrox/blob/1ab0f8534ea26b07d08ddb13b5759a9e2f80fda5/operator/Makefile#L400-L404

I can't see a way to add the static env var to a file without it getting overwritten by the generated output from the `operator-sdk` CLI.

File: image/templates/helm/stackrox-central/values.yaml.htpl (hunk 3 above)
Comment (kylape) [reply image/templates/helm/stackrox-central/values.yaml.htpl]: ah, it turns out that the config-shape.yaml file can't use the brackets because it's not a meta template (`.htpl`) file.

File: operator/api/v1alpha1/central_types.go (hunk 7 above)
Comment (kylape) [reply operator/api/v1alpha1/central_types.go:619]: Just to follow up here:  My preference would be to keep `enabled`, but if you would still prefer me to switch to `disable`, I am happy to do so.

File: image/templates/helm/stackrox-central/values.yaml.htpl (hunk 5 above)
Comment (kylape) [reply image/templates/helm/stackrox-central/values.yaml.htpl:544]: > keep installCRDs somewhat internal

After further consideration, I'd be willing to make this flag more internal.  Can you explain what that means in a bit more concrete terms?

**Comment by kylape** [comment]
I've thought about the CRD/Helm values config for a while, and here's where I'm at:  While I don't love adding a top level field called `configAsCode`, I am not sure there are any better options.  

* A section under `central.declarativeConfig` is confusing because 1) the config-controller is a separate deployment from Central, and 2) policy/config-as-code is a distinct feature from "Declarative Config"
* A different top-level name like `config-controller` is confusing because the options inside handle more than just the `config-controller` deployment.  There's the `disableCRD` flag, and the `enabled` field also toggles the `ROX_POLICY_AS_CODE` env var in the Central deployment.

Regarding the `installCRD` flag, I am still convinced there is value for customers to have it, though I admit it is likely an edgy edge case (i.e. quite rare for a customer to want to use it).  So I am open to moving the flag somewhere else more hidden, but I don't know where would be helpful.  Any suggestions are welcome!

File: operator/api/v1alpha1/central_types.go (hunk 7 above)
Comment (mclasmeier) [reply operator/api/v1alpha1/central_types.go:610]: Personally, I am not leaning to any side here. It's just that I believe we should (1) accept that we haven't been consistent in the past and (2) nevertheless strive for more consistency on our path forward. And the most recent example of an addition that I had in mind was the `scannerV4.disabled` toggle (there might be others). So that would be an indicator for switching to `disabled`-style toggles.

I would also be fine with bringing this up in some eng Slack channel in order to increase visibility on this matter. WDYT?

File: operator/bundle_helpers/patch-csv.py (hunk 9 above)
Comment (porridge) [reply operator/bundle_helpers/patch-csv.py:109]: I think [adding it here](https://github.com/stackrox/stackrox/blob/ed3a068a7a57052591972c10620529e931eab1d1/operator/config/manager/manager.yaml#L65) might do the trick?

Comment (porridge) [reply operator/bundle_helpers/patch-csv.py:109]: But if it's added unconditionally, then why do we need the env variable at all?

Comment (mclasmeier) [reply operator/bundle_helpers/patch-csv.py:109]: > I think [adding it here](https://github.com/stackrox/stackrox/blob/ed3a068a7a57052591972c10620529e931eab1d1/operator/config/manager/manager.yaml#L65) might do the trick?

Thank you, that's what I was looking for when I made these review remarks.

Comment (kylape) [reply operator/bundle_helpers/patch-csv.py:109]: Thanks for the pointer @porridge!

That said, we just agreed to remove the feature flag altogether in favor of the ability to toggle the feature via Central CR or Helm values file, which means we no longer need to make this change :tada:.

Comment (mclasmeier) [reply operator/bundle_helpers/patch-csv.py:109]: In that case let me resolve this.

File: operator/api/v1alpha1/central_types.go (hunk 7 above)
Comment (mclasmeier) [reply operator/api/v1alpha1/central_types.go:610]: Alright, don't treat this as a blocker from my side. Personally I would also be in favor of the positive form. Thank you for bringing it up on Slack -- looks like at least among those who participated in the thread, there is some tendency for the positive form as well.

**Comment by kylape** [comment]
Since https://github.com/stackrox/stackrox/pull/12966 merged, I've resolved all conversations about feature flags since it has been removed.  I think the only remaining open conversations are related to the shape of the CRD spec.  I currently have it as:

```
configAsCode:
  enabled: true
  installCRD: true
```

@mclasmeier If you don't mind, please summarize where you stand with this question, and I'd be happy to further address any further concerns.

**Comment by mclasmeier** [comment]
Hi, sorry.

So, regarding CRDs: This is basically the main point I am a little bit worried about, because this would be a new scenario (CRDs are somewhat different from the other manifests) within our Helm charts and it's not entirely clear to me how to get it right form the beginning.

Question, have you seen considered to use Helm's built-in mechanism for installing CRDs?

For example, https://helm.sh/docs/chart_best_practices/custom_resource_definitions/#method-1-let-helm-do-it-for-you:

> With the arrival of Helm 3, we removed the old crd-install hooks for a more simple methodology. There is now a special directory called crds that you can create in your chart to hold your CRDs. These CRDs are not templated, but will be installed by default when running a helm install for the chart. If the CRD already exists, it will be skipped with a warning. If you wish to skip the CRD installation step, you can pass the --skip-crds flag.

How does this sound? We could remove the templating wrapping from the CRD and just move it into the `crd/` directory. At the same time there would be a standard Helm command line parameter for disabling the installation of the CRD.

Since we are using the Helm libraries for rendering the manifests within the operator, I am quite sure we would have the same flexibility there as well (for controlling when they are part of the rendered manifest bundle).

**Comment by kylape** [comment]
The case for including the CRD in the template is found in this doc:

https://helm.sh/docs/chart_best_practices/custom_resource_definitions/#some-caveats-and-explanations

> There is no support at this time for upgrading or deleting CRDs using Helm.

The other big concern that typically concerns adding CRDs to the template is adding CRs of a CRD that's included in the template.  Having a separate `crds` folder ensures all CRDs are installed before any CRs are created.  In our case, we know at least in this release we won't include any `SecurityPolicy` CRs, so that's not a big concern (yet).

That said, I would like to validate exactly what is meant by "no support for upgrading CRDs".  Does this just mean changing the `version` for example from `v1alpha1` to `v1`, or does it mean any changes to the CRD's spec at all?  The former is much more acceptable than the latter.

While I'm thinking about it, we've had discussions about the `version` of our CRD.  I think we wanted to use `v1beta`.  Would you object to changing to that version, or should we maintain consistency with the other ACS CRDs with `v1alpha1`?

**Comment by kylape** [comment]
**tl;dr After some research, I am a bit more convinced keeping the CRD in the template is the best path forward for this release.**

After some testing, I determined that Helm takes the most conservative approach: it does not upgrade the CRD _at all_ after initial creation, even for adding a single field.

Given that, here are our options:

1. **Keep the CRD in the templates folder**.  CRD upgrades happen, but care must be taken to not cause data loss when modifying CRDs between releases.  If we ever want to add `SecurityPolicies` to the chart, this would be problematic.
2. **Move the CRD to the `crds` folder**.  This will force customers to manually upgrade CRDs.
3. **Move the CRD to a separate chart**.  More work to set up a separate chart, but should be able to auto-upgrade and avoids the issue where CRs fail to install because the CRD isn't yet installed/established.

I think the conservative approach taken with the `crds` folder option makes this pretty unusable to us.  We need the CRD to be automatically upgraded, and I assume **we take on the responsibility to ensure we don't cause data loss when upgrading a CRD**.

The separate helm chart is interesting, but it's not clear what it buys us.  And from the [helm doc](https://helm.sh/docs/chart_best_practices/custom_resource_definitions/#method-2-separate-charts) on it:

> each chart must be installed separately

Again, adding a manual step for the customer makes this a less desirable option.

That leaves us with keeping the CRD in the template folder.  There are a couple of disadvantages I can see of this approach:

1. Helm adds its labels to the CRD.  (It doesn't do this with the `crds` folder approach)
2. Clusters with more than one stackrox-central-services installation could see conflicts with the CRD, though the `installCRDs: false` option would work around this issue.
3. The operator must set `installCRDs: false`, though this is already addressed.

There is a possibility of wanting to add `SecurityPolicies` (or some other yet-to-be-created config-as-code CR) to the installation, but we can either work with the constraints established here (e.g. create a separate chart for default policies), or perhaps switch to putting the CRD in a separate chart in a future release (the main concern here would be transferring "ownership" of the CRD from one helm chart to another).

**Comment by kylape** [comment]
Another way to look at this choice is in terms of the desired installation experience we want for our users.

Do we want to have a "smooth" installation and upgrade experience where users trust us to make wise choices during installation, or do we want a more "hands on" approach where we trade some simplicity for control, thus delegating some risk (i.e. upgrading CRDs) to the user?

The "smooth" approach is including the CRD in the template, whereas the "hands on" approach is including the CRD in the `crds` folder.

File: operator/api/v1alpha1/central_types.go [hunk 12]
```diff
@@ -602,6 +606,10 @@ const (
...
```
Comment (porridge) [inline operator/api/v1alpha1/central_types.go:610]: Please change this to an enum, we're [trying to avoid booleans](https://github.com/stackrox/stackrox/blob/master/operator/EXTENDING_CRDS.md#data-type-choices).

**Comment by kylape** [comment]
So moving the CRD to the `crds` folder breaks manifest installation since it renders the helm chart by just looking at the templates folder (afaict).

**Comment by porridge** [comment]
> So moving the CRD to the `crds` folder breaks manifest installation since it renders the helm chart by just looking at the templates folder (afaict).

What a great opportunity to start dropping features from the manifest-based installation method 😉

**Comment by kylape** [comment]
I added a couple lines to the deploy script to apply the CRD, but it felt _too easy_ given that it was inside a bash library function.

**Comment by porridge** [comment]
FTR, as this PR still seems to be a in flux a bit, I'm still waiting for an explicit "go ahead" to start reviewing.

**Comment by kylape** [comment]
> I'm still waiting for an explicit "go ahead" to start reviewing.

I did another set of local validation, and everything seems to be working afaict.

After reviewing the code changes again, I just pushed a couple changes:

1. I correctly set the full path to the CRD in `deploy/common/k8sbased.sh`
2. I re-enabled the failing scanner v4 tests.

I feel like this PR is ready to review, unless you'd prefer to wait for CI to pass.  I'll check back in a couple hours to see if CI has passed.

File: operator/Makefile [hunk 13]
```diff
@@ -441,6 +441,7 @@ bundle-post-process: test-bundle-helpers operator-sdk ## Post-process CSV file t
 	mkdir -p build/
 	rm -rf build/bundle
 	cp -a bundle build/
+	cp ../config-controller/config/crd/bases/*.yaml build/bundle/manifests/
```
Comment (porridge) [inline operator/Makefile:444]: If possible, please list files by name, rather than use `*` here. It will make it easier to look for references and spot issues if ($DEITY forbid) there ever are clashes. Perhaps use `cp < /dev/null -iv` as well?

File: operator/api/v1alpha1/central_types.go [hunk 14]
```diff
@@ -75,6 +75,10 @@ type CentralSpec struct {
 	// Network configuration.
 	//+operator-sdk:csv:customresourcedefinitions:type=spec,displayName=Network,order=10,xDescriptors={"urn:alm:descriptor:com.tectonic.ui:advanced"}
 	Network *GlobalNetworkSpec `json:"network,omitempty"`
+
+	// Config-as-Code configuration
```
Comment (porridge) [inline operator/api/v1alpha1/central_types.go:79]:
```suggestion
	// Config-as-Code configuration.
```

File: operator/api/v1alpha1/central_types.go [hunk 15]
```diff
@@ -602,6 +606,24 @@ const (
 	ScannerV4ComponentDisabled ScannerV4ComponentPolicy = "Disabled"
 )
 
+type ConfigAsCodeSpec struct {
+	// If you want to deploy Config as Code components set this to "Enabled"
```
Comment (porridge) [inline operator/api/v1alpha1/central_types.go:610]:
```suggestion
	// If you want to deploy the Config as Code component set this to "Enabled".
```

File: operator/api/v1alpha1/central_types.go [hunk 16]
```diff
@@ -602,6 +606,24 @@ const (
...
+	//+kubebuilder:default=Enabled
+	//+operator-sdk:csv:customresourcedefinitions:type=spec,order=1,displayName="Config as Code component"
+	ComponentPolicy *ConfigAsCodeComponentPolicy `json:"configAsCodeComponent,omitempty"`
+}
+
+// ConfigAsCodeComponentPolicy is a type for values of spec.scannerV4.scannerComponent
```
Comment (porridge) [inline operator/api/v1alpha1/central_types.go:616]:
```suggestion
// ConfigAsCodeComponentPolicy is a type for values of spec.configAsCode.configAsCodeComponent.
```

File: operator/bundle_helpers/patch-csv.py [hunk 17]
```diff
@@ -91,6 +91,21 @@ def patch_csv(csv_doc, version, operator_image, first_version, related_images_mo
         # Ref https://osbs.readthedocs.io/en/latest/users.html?highlight=relatedImages#creating-the-relatedimages-section
         del csv_doc['spec']['relatedImages']
 
+    # Add SecurityPolicy CRD to ACS operator CSV
+    policy_crd = {
+        "name": "securitypolicies.config.stackrox.io",
+        "version": "v1alpha1",
+        "kind": "SecurityPolicy",
+        "displayName": "Security Policy",
+        "description": "SecurityPolicy is the Schema for the policies API",
```
Comment (porridge) [inline operator/bundle_helpers/patch-csv.py:100]:
```suggestion
        "description": "SecurityPolicy is the Schema for the policies API.",
```
I'm not a native speaker, but the capitalization of `Schema` and `policies` looks inconsistent..

File: operator/internal/values/translation/translation.go [hunk 18]
```diff
@@ -416,3 +416,12 @@ func GetGlobalNetwork(s *platform.GlobalNetworkSpec) *ValuesBuilder {
 	}
 	return &sv
 }
+
+// GetConfigAsCode converts *platform.GetConfigAsCodeSpec into *ValuesBuilder
```
Comment (porridge) [inline operator/internal/values/translation/translation.go:420]:
```suggestion
// GetConfigAsCode converts a *platform.GetConfigAsCodeSpec into a *ValuesBuilder.
```

**Review by porridge — APPROVED** [review]
There are plenty of nitpicks inline, but I'm going to tentatively approve given I'm out tomorrow, and freeze is on Monday. But please do follow up if you disagree on something.

Also, I don't really understand what this sentence in the PR description is trying to day:
> The name `configAsCode` is used to make the API future-proof in case we add other config-related CRDs down the road.

Ah, you mean "instead of `policyAsCode`"? +1 to that, but please mention this.

Also, it might make sense to add some more content to the description of the struct so it's obvious in the operator UI what this is about. TBH I wouldn't know if I didn't have the background.

And the last thing, please try to make sure this renders OK in the OpenShift console!

Thanks for re-enabling the test!

File: Makefile [hunk 19]
```diff
@@ -362,8 +362,7 @@ clean-proto-generated-srcs:
 config-controller-gen:
 	make -C config-controller/ manifests
 	make -C config-controller/ generate
-	echo -e '{{- include "srox.init" . -}}' > image/templates/helm/stackrox-central/templates/00-securitypolicies-crd.yaml
-	cat config-controller/config/crd/bases/config.stackrox.io_securitypolicies.yaml >> image/templates/helm/stackrox-central/templates/00-securitypolicies-crd.yaml
+	cp config-controller/config/crd/bases/config.stackrox.io_securitypolicies.yaml image/templates/helm/stackrox-central/crds
```
[the remark this answers is not in the document]
Comment (kylape) [reply Makefile:365]: [Yes](https://github.com/stackrox/stackrox/blob/d16183ca67600dc42bed4a79aeb8f520431c0a8f/scripts/ci/jobs/check-generated.sh#L66-L78)

File: operator/Makefile (hunk 13 above)
Comment (kylape) [reply operator/Makefile:444]: I'm not sure `-i` is helpful, but happy to discuss in a follow up.

File: operator/bundle_helpers/patch-csv.py (hunk 17 above)
Comment (kylape) [reply operator/bundle_helpers/patch-csv.py:100]: No you're right, capitalizing schema doesn't make sense.

File: image/templates/helm/stackrox-central/templates/02-config-controller-02-deployment.yaml
[the remark this answers is not in the document]
Comment (kylape) [reply image/templates/helm/stackrox-central/templates/02-config-controller-02-deployment.yaml:1]: https://issues.redhat.com/browse/ROX-26749

File: deploy/common/k8sbased.sh [hunk 20]
```diff
@@ -151,6 +151,11 @@ function launch_central {
         prompt_if_central_exists "${central_namespace}"
     fi
 
+    echo "Installing SecurityPolicy CRD..."
+
+    crd_path=$(realpath "$(git rev-parse --show-toplevel)/config-controller/config/crd/bases/config.stackrox.io_securitypolicies.yaml")
+    kubectl apply -f $crd_path
```
[the remark this answers is not in the document]
Comment (kylape) [reply deploy/common/k8sbased.sh]: Sent a follow up message in slack.

File: operator/Makefile (hunk 13 above)
Comment (porridge) [reply operator/Makefile]: I hoped `-i` would prevent overwriting a file present in `bundle/manifests/` in case another file with the same name ever appeared in `../config-controller/config/crd/bases` @kylape .
